[CORE-9797] schema_registry: Improve diagnostics when references are missing

BenPope opened the pull request:
Improve diagnostics when references are missing during:
* `POST /subjects/{subject}/`
* `POST /subjects/{subject}/versions`
* `POST /compatibility/subjects/{subject}/versions`

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Improvements

* Schema Registry: Improve diagnostics when POST a schema with missing references

BenPope committed:
sr/json: Missing references: Add ducktape test

Test that missing references return an error message with the
correct details.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr/proto: Missing references: Check references

Produce the correct error response when a reference is missing.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr: Missing references: Render schema in error message

This fixes a regression in #21323, the schema should be in the
error response message.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr/avro: Missing references: Check references

Note: This is intentionally not tested in ducktape as
collect_references wasn't previously called, and this is a breaking
change.

There will be an improvement in the error message if a compatibility
check has missing reference.

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr/avro: Support normalizing references

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -463,14 +473,7 @@ ss::future<pb::FileDescriptorProto> import_schema(
-        co_return co_await build_file_with_refs(
-          dp, store, schema.share(), norm);
-    } catch (const exception& e) {
-        vlog(
-          plog.warn, "Failed to decode schema {}: {}", schema.sub(), e.what());
-        throw as_exception(invalid_schema(schema));
```
This is actually important to prevent a lot of noise in the response message
```suggestion
        // Rethrow if the schema is missing references
        if (e.code() == error_code::schema_empty) {
            throw;
        }
        // Otherwise log the error details and throw an appropriate error for
        // the response
        vlog(
          plog.warn, "Failed to decode schema {}: {}", schema.sub(), e.what());
        throw as_exception(invalid_schema(schema));

pgellert commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -427,13 +427,15 @@ post_subject(server::request_t rq, server::reply_t rp) {
           std::move(unparsed.def), norm);
     } catch (const exception& e) {
-        if (e.code() == error_code::schema_empty) {
+        if (e.code() == error_code::schema_empty && len == 0) {
```
The ducktape test makes sense but I'm not following this part. Why is this part necessary?

BenPope replied:
>  Why is this part necessary?
The `error_code` for an empty schema (which can come from avro) is shared with `no_reference_found_for`, so this change allows both errors to be distinguished.

I'll rename it to `schema_empty_or_missing_reference`, and add some helpers.

BenPope replied:
I added a new `error_code`

pgellert commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -467,6 +477,12 @@ ss::future<pb::FileDescriptorProto> import_schema(
         co_return co_await build_file_with_refs(
           dp, store, schema.share(), norm);
     } catch (const exception& e) {
+        // Rethrow if the schema is missing references
+        if (e.code() == error_code::schema_empty) {
```
Shouldn't this be `no_reference_found_for` based on the comment? But also, could you elaborate while this special casing is needed? Wouldn't we want to log either way and wrap the exception?

BenPope replied:
> Wouldn't we want to log either way and wrap the exception?

The error response has sufficient detail, and is logged elsewhere.

pgellert commented on src/v/pandaproxy/schema_registry/avro.cc:
```diff
@@ -560,11 +560,16 @@ ss::future<collected_schema> collect_schema(
+                  ref.sub, ref.version, include_deleted::yes);
+                collected = co_await collect_schema(
+                  store, std::move(collected), ref.name, std::move(ss.schema));
+            } catch (const exception& e) {
+                throw as_exception(
+                  no_reference_found_for(schema, ref.sub, ref.version));
+            }
```
Should we check the error code here as well?
```
if (
              e.code() == error_code::subject_not_found
              || e.code() == error_code::subject_version_not_found) {
```

BenPope replied:
There are a few of these, I'll add a helper.

BenPope replied:
Done

pgellert commented on src/v/pandaproxy/schema_registry/avro.cc:
```diff
@@ -636,6 +641,16 @@ sanitize_avro_schema_definition(schema_definition def) {
+  schema_getter& store, subject_schema unparsed_schema, normalize) {
+    auto [sub, schema] = std::move(unparsed_schema).destructure();
+    // Check references
+    co_await collect_schema(store, {}, sub, {sub, schema.share()});
```
This seems to be new behaviour. Would it make sense to extract this to a separate PR and backport everything else but this?

BenPope replied:
> This seems to be new behaviour. Would it make sense to extract this to a separate PR and backport everything else but this?

I've made it a TODO, it improves the error message when a compatibility check is required.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d162ff49-093a-459b-8f67-4c94f44e95ce-->
<!--TEST RESULTS-->
<!--026aa899-cce0-4f5d-9e97-5163d828c7c6-->
<!--TEST RESULTS-->
<!--f84f6632-c242-4879-8512-6ac29669fb20-->
<!--TEST RESULTS-->
<!--65e13223-0b7c-472c-b1e0-0850199d47d6-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64548">build#64548</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.archival_test.ArchivalTest.test_all_partitions_leadership_transfer.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64548#019639c5-62ca-4741-af31-de1386786218 | FLAKY | 1/2 |
| rptest.tests.cloud_storage_scrubber_test.CloudStorageScrubberTest.test_scrubber.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64548#019639c2-7805-44ca-90bb-43eed9903f35 | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/64548#019639c2-7804-41a0-a47c-ba4a8aa74d55 | FLAKY | 1/2 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64548#019639c2-7806-44aa-bfe9-c849bc84ef58 | FLAKY | 1/3 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64624">build#64624</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.rest_catalog_connection_test.RestCatalogConnectionTest.test_redpanda_connection_to_rest_catalog.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/64624#01963fd6-b540-4601-9f9f-fad9452eed07 | FLAKY | 20/21 |
| rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=node_add.test_mode=TestMode.TIRED_STORAGE.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/64624#01963fd6-b545-41a4-b68f-8bdc13ca1eb9 | FLAKY | 16/17 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/64624#01963fd6-b540-4867-a22b-1dbf99109a89 | FLAKY | 15/21 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/64624#01963fd6-9d10-4fa5-ae62-9ade0e9a6272 | FLAKY | 18/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64678">build#64678</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64678#01964423-950f-49e2-b0bc-d591653c2c56 | FLAKY | 20/21 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64678#01964423-9510-4446-8507-cb7568f62548 | FLAKY | 11/21 |
| rptest.tests.datalake.throttling_test.DatalakeThrottlingTest.test_basic_throttling.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/64678#01964427-8bea-441e-bde8-9c751d8a9820 | FLAKY | 17/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/64678#01964423-9510-4446-8507-cb7568f62548 | FLAKY | 8/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/64838">build#64838</a></summary>


| test_id                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit        | https://buildkite.com/redpanda/redpanda/builds/64838#01965c7c-7e55-43ed-86bb-edea76a87757 | FLAKY         | 1/2      |
</details>
<!--/body-->

BenPope committed:
sr: Refactor: Split error_code::schema_empty

No functional change.

The error code is also used when a schema is missing references

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr: Refactor: Introduce failed_subject_schema_lookup

No functional change

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
sr: Ensure logging is on a single line

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/2e3fb38ee91f55639ff7d16381fe2dd2ae38a04a..004ece030171d2d6546dc6ed1ab851c626f898b1)
* Split `error_code::schema_empty`
* Add a helper for `failed_subject_schema_lookup`
* Add an `operator<<` for `raw_string`
* Don't force a failure for avro references

BenPope commented:
Changees in [force-push](https://github.com/redpanda-data/redpanda/compare/3b871daa0cd5ae32e9b883215e4406cb29487558..f771174d77adfe8457cd0604a464b316aeb619d0)
* Rebase to fix logging conflicts

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64624-->
<!--739dc4e3-6903-4ae8-824d-818961ddb9df-->
<!--d2d1bb3c-b6d2-42ef-b125-2a3064b5dcfc-->
<!--title-->
## Retry command for [Build#64624](https://buildkite.com/redpanda/redpanda/builds/64624)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/transactions/producers_api_test.py::ProducersAdminAPITest.test_producers_state_api_during_load
tests/rptest/tests/datalake/rest_catalog_connection_test.py::RestCatalogConnectionTest.test_redpanda_connection_to_rest_catalog@{"cloud_storage_type":1}
```
<!--/body-->

pgellert commented on src/v/pandaproxy/schema_registry/error.cc:
```diff
@@ -112,6 +114,7 @@ struct error_category final : std::error_category {
         case error_code::schema_invalid:
             return reply_error_code::unprocessable_entity;
         case error_code::schema_empty:
+        case error_code::schema_missing_reference:
```
Nice, looks great

pgellert commented on src/v/pandaproxy/schema_registry/sharded_store.cc:
```diff
@@ -897,7 +897,7 @@ ss::future<compatibility_result> sharded_store::do_is_compatible(
               fmt::format("{{oldSchemaVersion: {}}}", old_schema.version));
             version_messages.emplace_back(
-              fmt::format("{{oldSchema: '{}'}}", to_string(old_valid.raw())));
+              fmt::format("{{oldSchema: '{}'}}", old_valid.raw()));
```
The commit message says:
> Change operator<< so that newlines are escaped.

This corresponds to the last commit, right? I think for that, I don't think it matters whether you stream the output or format a string. Am I missing something here? (Mostly just checking that I'm not missing anything here + that you didn't intend to change anything more here.)

BenPope replied:
You are correct, I had previously implemented operator<< in terms of a `fmt::formatter`, that was able to escape characters, but the final commit is simpler.

I'll drop this commit and use `iobuf_const_parser::read_string` in the next, so that utf-8 validation occurs.

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/f771174d77adfe8457cd0604a464b316aeb619d0..48d615e82cd007b059c0922dce008bf5ba500f3e)
* Drop the commit implementing `operator<<` for `raw_string`
* Replace the use with `iobuf_const_parser::read_string`

IoannisRP commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -437,13 +437,21 @@ ss::future<pb::FileDescriptorProto> build_file_with_refs(
+              subject_schema{subject{ref.name}, std::move(dep.schema).def()},
+              normalize::no);
+        } catch (const exception& e) {
+            if (failed_subject_schema_lookup(e.code())) {
+                throw as_exception(
+                  no_reference_found_for(schema, ref.sub, ref.version));
+            }
```
This can fail for 2 reasons.
- Schema `{sub, version}` not found
- Ref schema failing to build.
In both cases we return the same error that the schema cannot be found.

I think it would be better if we had a different error for each type of reference failure.

It would be nice to also have some recursive info on where the failure occured, in case it happened further in the dep chain, but I don't think it is worth the effort.

This affects only `proto`. `avro` and `json` don't have any processing once the schema is retrieved.

BenPope replied:
I don't understand your reasoning.

If it fails due to a reference not being found, it will translate to a `no_reference_found_for`.
If it fails because a reference failed to build, that would be returned as the original failure.

IoannisRP replied:
You are right. I misread this.
Ignore my comment

IoannisRP reviewed (dismissed):
lgtm

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/48d615e82cd007b059c0922dce008bf5ba500f3e..db608e1f8292d79f6fce10aee08c50d1a0bc4c0a)
* Rebase to avoid logging conflict in `reply.h`

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/64678-->
<!--0691ad8e-42df-4ec2-a987-adaf5f6064ba-->
<!--6b3bfb8b-4fac-479b-8d8d-7e012349dfd1-->
<!--b04a8117-cc05-4817-a0d4-eba2b20a8020-->
<!--title-->
## Retry command for [Build#64678](https://buildkite.com/redpanda/redpanda/builds/64678)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":false},"transfer_leadership":false}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":false},"transfer_leadership":true}
tests/rptest/tests/upgrade_test.py::UpgradeBackToBackTest.test_upgrade_with_all_workloads@{"single_upgrade":false}
tests/rptest/tests/datalake/throttling_test.py::DatalakeThrottlingTest.test_basic_throttling@{"catalog_type":"nessie","cloud_storage_type":1}
```
<!--/body-->

BenPope commented:
/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":false},"transfer_leadership":false}
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_migrated_topic_data_integrity@{"params":{"cancellation":{"dir":"in","stage":"preparing"},"use_alias":false},"transfer_leadership":true}
tests/rptest/tests/upgrade_test.py::UpgradeBackToBackTest.test_upgrade_with_all_workloads@{"single_upgrade":false}
tests/rptest/tests/datalake/throttling_test.py::DatalakeThrottlingTest.test_basic_throttling@{"catalog_type":"nessie","cloud_storage_type":1}

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x
